src/hypothesis/searchstrategy/strategies.py

```diff
@@ -140,7 +142,17 @@ def recur(strat):
             else:
                 needs_update = None
 
+            count = 0
+            seen = set()
             while needs_update:
+                count += 1
+                # If we seem to be taking a really long time to stabilize we
+                # start tracking seen values to attempt to detect an infinite
+                # loop.
+                if count > 50:
+                    key = frozenset(mapping.items())
+                    assert key not in seen, (key, name)
```

Comment:
This should raise an exception (probably AssertionError or InvalidState), possibly with a useful message for users, because I can't see where we'd catch it.  Also: running Python with `-O` strips out `assert` statements, so don't use them for control flow!

Reply:
No, this is a normal assertion. This condition is supposed to be impossible.